paperqa/docs.py

```diff
@@ -412,23 +412,20 @@ async def aadd_texts(
         Returns:
             True if the doc was added, otherwise False if already in the collection.
         """
-        all_settings = get_settings(settings)
-
-        if embedding_model is None:
```

Comment:
Why cut the option to pull the model from settings here? I think we can still support it and skip if the chunksize = 0.

Reply:
I switched to embed later - if embedding_model isn't present, we will no longer fetch from settings. That means the settings isn't used anywhere in the code.

Reply:
Right, I'm saying that instead of the flag being the presence of `embedding_model` in the entrypoint, what do you think of using chunksize=0 from settings? Then that way a user could still use this method just by inputting a settings object and they don't need to instantiate their embedding_model before passing.

Reply:
Chunk_size / skipping chunking does come from settings.

The change I made here is unrelated to this. I just made it so that embedding is not done when a document is added, since it's kind of wasteful for tests and rate limits. Now we embed texts only when calling gather evidence. A side-effect of that change is the only time we embed is if an explicit embedding_model is passed

Reply:
I see -- I was imagining that a user would only want to avoid the early embedding if they wanted to do full-doc embeddings.

One other edge case, when a user is building an index, we use `aadd`/`aadd_texts` for each doc as we index it. There we probably want to pre-embed because we save individual `Docs` with one paper each, and combine the candidate matching `Docs` after a paper search. If we only embedded at gather evidence time, then we'd need to either embed each time, or split back out into the constituent single-paper `Docs` objects and re-save them at that step. 

To your point that definitely can push the embedding rate limit, but I think it makes sense for the local index. This PR totally supports this use case if you just insert an `embedding_model`, but in practice we're only  passing a `Settings` object into our `aadd` method call in `search.process_file`. So it was clean to extract the model out of the settings there.

I think all we need to do instead is to build an `embedding_model` in `search.process_file` and pass that into `aadd` and then this will still work.

Reply:
OK - added new setting (off by default) to defer_embeddings. It is default on for debug setting (to avoid so many embedding calls) and if we get to a point where we have extremely large indices, it may be more efficient to defer embedding.